Code Refactor and Reprocessing Functionality

LukeSeifert opened the pull request:
This PR includes a refactor of the code from the `flow-tests` directory, proper incorporation of Serpent2 built-in flow functionality, and several different approximations.

The `Notes` directory contains a Jupyter notebook with information on how the different reprocessing constant approximations are generated.

The `msbr_compare` directory is where the refactor and new functionality is located. The `readme` contains most of the information on the overall purpose of the scripts, as well as what each individual script is meant to do.

pep8speaks commented:
Hello @LukeSeifert! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`msr-flows/msbr_compare/misc_funcs.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/misc_funcs.py):

> [Line 53:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/misc_funcs.py#L53): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)

* In the file [`msr-flows/msbr_compare/run_user_input.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py):

> [Line 1:1](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L1): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 14:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L14): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 36:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L36): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 46:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L46): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 48:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L48): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 113:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L113): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 120:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L120): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 79 characters)
> [Line 122:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L122): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 124:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L124): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 216:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L216): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 223:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L223): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 79 characters)
> [Line 225:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L225): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 227:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L227): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 412:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L412): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 429:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L429): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 444:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L444): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 462:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L462): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 477:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/run_user_input.py#L477): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

* In the file [`msr-flows/msbr_compare/saltproc_read.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py):

> [Line 146:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 79 characters)
> [Line 158:5](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py#L158): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 159:5](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py#L159): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 173:5](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py#L173): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 209:5](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/saltproc_read.py#L209): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

* In the file [`msr-flows/msbr_compare/serpent_calculations.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py):

> [Line 8:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L8): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 79 characters)
> [Line 10:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L10): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (170 > 79 characters)
> [Line 61:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L61): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 77:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L77): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 79 characters)
> [Line 96:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 79 characters)
> [Line 98:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 153:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L153): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 170:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 179:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L179): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 213:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L213): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 231:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L231): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 253:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_calculations.py#L253): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)

* In the file [`msr-flows/msbr_compare/serpent_input.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_input.py):

> [Line 34:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_input.py#L34): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 114:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_input.py#L114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 212:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_input.py#L212): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)

* In the file [`msr-flows/msbr_compare/serpent_output.py`](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py):

> [Line 17:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 40:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 78:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L78): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 124:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L124): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 79 characters)
> [Line 125:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L125): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 79 characters)
> [Line 143:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L143): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 191:80](https://github.com/arfc/msr-neutronics/blob/c4e0efbba82722842efcd618a43cb737f8b66904/msr-flows/msbr_compare/serpent_output.py#L191): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+   "source": [
+    "## Linear Growth Approximation\n",
+    "\n",
+    "#### Only suitable for SS\n",
```
Use expanded acronym on first use.
```suggestion
    "#### Only suitable for steady state (SS) calculations \n",
```

yardasol replied:
tagging @nsryan2 for another opinion

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "#### Only suitable for SS\n",
+    "\n",
+    "The linear growth approximation is an approach which assumes that over some time step X, the growth is approximately linear. This means that the removal constant must be calculated such that the number of atoms at time X is the same as SaltProc at that time. As long as accuracy can be maintained to some position past where the SaltProc value is determined, then this method can be used to red ...
```
Here's some formatting and flow suggestions to enhance clarity. I'm not quite sure if my edits here reflect what you're trying to convey. If they don't, you can ignore this suggestion.
```suggestion
    "The _linear growth approximation_ is an approach which assumes that over some time step X, the growth is approximately linear. This means that the removal constant must be calculated such that the number of atoms at time X is the same as in the SaltProc calculation at time X. As long as accuracy can be maintained to some position past where the SaltProc value is determined, then this method can be used to reduce the computational time of SaltProc by making a series of linear approximations.\n",
```

yardasol replied:
tagging @nsryan2 for another opinion

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "The linear growth approximation is an approach which assumes that over some time step X, the growth is approximately linear. This means that the removal constant must be calculated such that the number of atoms at time X is the same as SaltProc at that time. As long as accuracy can be maintained to some position past where the SaltProc value is determined, then this method can be used to red ...
+    "\n",
+    "![title](LGA.png)\n",
```
This is a great figure for illustrating what the LGA does! Good job.

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "![title](LGA.png)\n",
+    "\n",
+    "This approach has 3 main steps\n",
```
Good job explaining the steps of this algorithm!

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "    $$C = \\frac{N_f - N_i}{t_f - t_i}$$\n",
+    "\n",
+    "2. Plug the constant growth value in and then include reprocessing with the condition that it must match SaltProc at a given time. This is done by calculating waste atoms and solving with scipy.optimize root function.     \n",
```
Flow + formatting edit for enhanced clarity. Same quantifier on your intended message applies to this suggestion.
```suggestion
    "2. Plug the linear growth term into the Bateman equation and then include reprocessing with the condition that it must match the SaltProc value at a given time. This is done by calculating waste atoms and solving with `scipy.optimize` root function.     \n",
```

yardasol replied:
tagging @nsryan2 for another opinion

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+   "source": [
+    "## Cycle Time Decay Approximation\n",
+    "\n",
+    "#### Suitable for BOC/SS/EOC\n",
```
```suggestion
    "#### Suitable for all calculations\n",
```

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "This approximation treats the type 1 flow as a decay (which is how it is viewed mathematically). This means that the \"decay\" constant can be calculated from the \"half-life\" of a particular target.\n",
+    "\n",
+    "In this case, the half-lives are determined from the cycle times from Robertson et al. report on the MSBR that Andrei cites in the MSBR example.\n",
```
>the half-lives are determined from the cycle times from Robertson et al. report on the MSBR that Andrei cites in the MSBR example.

This is a bit vague. Consider using the path to the file where the citation to the Robertson paper exists. Or if possible, include a link to the paper directly.

yardasol commented on msr-flows/Notes/type_1_flow.ipynb:
```diff
@@ -0,0 +1,152 @@
+    "\n",
+    "Since some elements may not have isotopes which are necessarily important enough to track that closely (for neutronics purposes), we will set any element without important isotopes to use the base Linear Growth Approximation.\n",
+    "\n",
+    "https://www.osti.gov/etdeweb/servlets/purl/20332542 \n",
```
It's not clear to me what this link is for. Perhaps you could user hypertext to clarify your intent?

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+## Overview
+
+These scripts are designed to generate a Serpent2 input deck, run the deck, and then generate plots. Currently, the MSBR is analyzed with SaltProc and then compared with different approximations using Serpent2's built-in reprocessing functionality. 
```
excellent description

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### User Input
+Takes all run settings in a single location and pass them along.
+
+### Run User Input
```
```suggestion
### `run_user_input.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+
+## Explanation of Scripts
+
+### User Input
```
```suggestion
### `user_input.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### Run User Input
+Defines the specific run cases to be used. Takes the user input data and the specific run case functions and uses them to generate output.
+
+### Serpent Input
```
```suggestion
### `serpent_input.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### Serpent Input
+Builds the Serpent2 input deck based on settings provided. Runs the Serpent input deck and checks that it has run successfully.
+
+### Serpent Calculations
```
```suggestion
### `serpent_calculations.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### Serpent Calculations
+Provides calculations for reprocessing constants of different approximations.
+
+### Serpent Output
```
```suggestion
### `serpent_output.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### Serpent Output
+Deals with the depletion and results outputs Serpent2 provides. Also parses SaltProc material files.
+
+### Miscellaneous Functions
```
```suggestion
### `misc_funcs.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+### Miscellaneous Functions
+Contains one-time-use functions or minor functions which don't necessarily fit in one of the other major categories.
+
+### SaltProc Read
```
```suggestion
### `saltproc_read.py`
```

yardasol commented on msr-flows/msbr_compare/readme.md:
```diff
@@ -0,0 +1,51 @@
+- Using ENDF instead of Jeff
+- The lambda terms of core flow assume constant density (core and piping same densities)
+
+## Things to do:
```
Consider making these issues in this repo

yardasol commented on msr-flows/msbr_compare/saltproc_read.py:
```diff
@@ -0,0 +1,223 @@
+
+def get_library_isotopes(acelib_path):
+    """ Returns the isotopes in the cross section library
+    Parameters:
```
Consider adding description for `acelib_path`

yardasol commented on msr-flows/msbr_compare/saltproc_read.py:
```diff
@@ -0,0 +1,223 @@
+from pyne import nucname
+
+
+def check_isotope_in_library(isotope, lib_isos):
```
consider adding `Paremeters` section to describe `isotope` and `lib_isos`

yardasol commented on msr-flows/msbr_compare/serpent_output.py:
```diff
@@ -0,0 +1,233 @@
+import serpentTools as st
```
I think `pyne` also has a serpent module. But you might have already tried using that.

LukeSeifert replied:
I wasn't aware of that actually, I'm just used to using Serpent Tools. I might have to look into the `pyne` version.

yardasol commented on msr-flows/readme.md:
```diff
@@ -1,50 +1,11 @@
-### Short Term Testing
+## Flow Tests
 
-Testing to see the short term effects cycle time has upon msr behaviour,
```
I assume these deletions are to reflect the refactor?

LukeSeifert replied:
Yeah, I was using the original scripts for testing different things, so the refactor also trimmed some of those.

yardasol replied:
Okay. Marking as resolved

yardasol reviewed (changes requested):
Hi @LukeSeifert,
Thanks for your PR! This repo hasn't been updated in a while so I'm glad to see you're giving it the TLC it deserves.

## General feedback

Overall, I really like your level of detail and documentation. There are some changes that I suggest you consider making before I approve this PR.

### PR needs linked issues
In the [page on making PR's on our website](https://arfc.github.io/manual/guides/pull_requests):
> PRs made in the ARFC organization should:
>   
  .
  .
  .
>    - Link to an issue

During the process of writing this review, I've looked at the description of this PR several times and had questions about what the described solutions are fixing and why this PR contains the things it does (e.g. What about the previous incorporation of Serpent2 build-in flow functionality was improper? Why does the code in the `flow-tests` directory need refactoring?). Knowing what issues this PR is resolving would help me understand the PR description. I've skimmed the existing issues in this repo and it looks like you probably have ones you can link here that would answer those questions.

### Merging from `master`
This may not be relevant but I'll say it anyways just in case it turns out to be. I see you are merging from your fork's `master` branch. While this repository doesn't seem to have any contributing guidelines, but if we take a look at [Moltres' `CONTRIBUTING.md` file](https://github.com/arfc/moltres/blob/devel/CONTRIBUTING.md):
>Keep your own "master" and "devel" branches in sync with the main repository's "master" and "devel" branches. Specifically, do not push your own commits directly to your "master" and "devel" branches.


My interpretation of this is that you can always create branches off of your forks `master` with the most recent working version of the code to develop multiple features independently (hopefully) without conflicting. I'm not clear on whether or not this general workflow applies to _all_ repos in this group, but regardless I thought I'd point this out just as something for you to consider.

## Review comments:
In addition to my feedback in line:
- Looks like you have some PEP8 issues. You can easily fix these by executing the following in the directory where you have your `.py` files:
   ```
   autopep8 --in-place *.py
   git add *.py
   git commit -m "pep8 fixes"
   git push
   ```
   rinse and repeat till @pep8speaks stops talking to you.
- `msr-flows/Notes/.ipynb_checkpoints` is cruft and should be removed from this PR
- There is some inconsistency with how you write docstrings in your `.py` files. For example, in the `_init_` function for the `full_run_serp` class in `run_user_input.py`, you have the following sections
   ```
   Returns
   -------
   None
   ```
   But later in `linear_generation_reprocessing_constants`, you have
   ```
   Returns
   -------
   
   ```
   and then in `cycle_time_decay`, there isn't a `Returns` section in the docstring at all. Ideally, all of your function docstrings would have a `Returns` section with a full description of the returned values, or `None` if no values are returned. Similar inconsistencies exist in the use of the `Parameters` section, and in the use of docstrings at all. Consider going back through your docstrings in all of your `.py` files and make the necessary edits to ensure they're all consistent.


## Reviewer checkist
- [x] Read the PR description
- [x] Read through all the changes, considering the following questions.
  - [x] Is there anything you can praise about this PR? Start with
        praise.
  - [x] Are variable names brief but descriptive?
  - [ ] Are new/changed functions no longer than a paragraph?
  - [ ] Do all function parameters have default values where appropriate?
  - [ ] Is the code clear and clean? (see Robert C. Martin's
        [Clean Code](https://i-share-uiu.primo.exlibrisgroup.com/discovery/fulldisplay?docid=alma99944155312205899&context=L&vid=01CARLI_UIU:CARLI_UIU&tab=LibraryCatalog&lang=en))
  - [x] Is there enough documentation?
  - [ ] Does the programming style meet the requirements of the
        repository ([PEP8](https://www.python.org/dev/peps/pep-0008/) for python,
        [google C++ style guide](https://google.github.io/styleguide/cppguide.html), etc.)
  - [ ] If there is no continuous integration, check out the branch
        locally, build, and run the tests.
  - [ ] Do the tests pass on your machine?
  - [ ] Is the PR free of random cruft (built files, `.swp` files,
        etc.)?
  - [ ] Do all files in the PR belong in the repository?
  - [ ] If the PR deletes files, is this appropriate?
  - [ ] If the PR adds files or data, are these new files compatible
        with the repository license?
  - [x] Make a review, leaving kind comments and suggesting changes
        where needed (to resolve the above).
  - [ ] Has the author resolved all of the comments and suggestions
        in your review?
- [ ] When you approve of the PR, merge and close it.
- [ ] Does this PR close an issue? If so, be sure to descriptively
      close this issue when the PR is merged
- [x] Thank the author for their contribution.

LukeSeifert commented:
@yardasol Thanks for the reviewing the PR! I'm pretty busy this week, but I should be able to get around to your comments next week or the week after.

LukeSeifert commented:
I originally had this work on its own branch before I merged it with master, that makes sense to PR the branch I was working on. I don't think I'm able to change it without closing this PR and opening a new one, but I will definitely keep it in mind for next time. 

I believe I addressed most of the suggestions you made, but the pep8 issues weren't fixable with autopep8 for some reason (I think because a lot of them were doc strings), so I did those manually. For the `saltproc_read.py`, since its a script I copied from SaltProc just for ease of access, I wasn't sure if I should make changes to the built-in functions, that's why I just added my own function to it and didn't change anything else. In the future, I should be able to structure things slightly differently such that I won't need to use that script anymore though, so I think it is only a temporary issue.

yardasol commented:
Looks like all your pep8 issues have been resolved. Nice work.
You should really add docstrings for the parameters in `get_library_isotopes` and `check_isotope_in_library`. Other than that, everything else looks good.

yardasol commented on msr-flows/msbr_compare/serpent_input.py:
```diff
@@ -0,0 +1,375 @@
+
+        Parameters
+        ----------
+        reprocessing : dictionary
```
You should add what the keys and values for the dictionary like this:

```
        reprocessing : dict
            Dictionary that...

            ``key``
                key description
            ``value``
                value description

```
do this for all `dict` parameters in your function docstrings.

yardasol reviewed (changes requested):
Hi @LukeSeifert, I've looked at your changes and I like how thorough you were with the additions to the docstrings.

I noticed a few typos that have suggestions ready for commit.

I brought up a few points in the in-line comments that I think are worth repeating here, in brief at least:
- Every docstring is now bounded by a `'''` instead of `"""`. `"""` is the correct format. Is there a reason you changed this?
- Make sure to add `key` and `value` descriptions to parameters of type `dict` in your docstrings. I've given you an example of how to do this in the in-line comment.

Sorry for being such a stickler about this. I think it'll save time in the long run to have these smaller pieces taken care of sooner rather than later (and doing these fixes now preps you for making those additions to your coding workflow in the future!)

LukeSeifert commented:
Thanks for the comments! I did not realize that the triple quote is used over the triple apostrophe, so I went through and replaced that in the other files. I also added the proper dictionary docstrings and fixed the typos you pointed out.

LukeSeifert commented on msr-flows/msbr_compare/serpent_calculations.py:
```diff
@@ -0,0 +1,277 @@
+
+    def repr_cnst_calc(self, iso_dict=False):
+        '''
+        Calculate reprocessing constants to be used for each element.
```
I think this should be alright as it is since there is some more context in the rest of the docstring now that I fixed the dictionary docstrings.

yardasol reviewed (approved):
Good work. I didn't see anything else looking over the code, so I think you can go ahead an merge!
